### PR TITLE
[FEAT] Add fugw interface

### PR DESCRIPTION
**Pull request by pbarbarant**
Solves #77 .

Added:

- [x] Fugw dense wrapper
- [x] Coarse-to-fine wrapper
- [x] Unit tests
- [x] Docstrings
- [x] Documentation

### CODE REVIEW

File: fmralign/alignment_methods.py [hunk 1]
```diff
@@ -731,3 +739,256 @@ def transform(self, X, verbose=False):
...
+            segmentation,
+            embeddings=geometry_embedding,
+            n_samples=n_samples,
+        )
+
+    def fit(
+        self,
+        X,
+        Y,
+        segmentation,
```
Comment (bthirion) [inline fmralign/alignment_methods.py:829]: sklearn types estimators tend to mvoe all parameters  (beyond X,Y) to the builder and have a very homgeneous fit(X,Y) API. Could we at least get closer to it ?

File: fmralign/alignment_methods.py [hunk 2]
```diff
@@ -731,3 +739,256 @@ def transform(self, X, verbose=False):
...
+            )
+
+            self.mapping = fine_mapping
+
+        return self
+
+    def transform(
+        self,
+        X,
+        device="auto",
```
Comment (bthirion) [inline fmralign/alignment_methods.py:955]: same point here.

**Review by bthirion — COMMENTED** [review]
LGTM otherwise.

**Comment by pbarbarant** [comment]
> Can you fix the CI ?

Done!

File: fmralign/alignment_methods.py (hunk 1 above)
Comment (bthirion) [reply fmralign/alignment_methods.py:823]: We should have a discussion on the API...

File: fmralign/alignment_methods.py [hunk 3]
```diff
@@ -731,3 +739,255 @@ def transform(self, X, verbose=False):
...
+        id_reg=0.0,
+        device="auto",
+    ):
+        """Project features using the fitted FUGW alignment
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_features)
+            Source features
+        id_reg: float, in the [0, 1] interval, defaults to 0
```
Comment (bthirion) [inline fmralign/alignment_methods.py:963]: these 2 parameters should be specified at the class creation step I think.

**Review by bthirion — COMMENTED** [review]
LGTM overall, but we need to agree on the API.

File: fmralign/alignment_methods.py (hunk 3 above)
Comment (pbarbarant) [reply fmralign/alignment_methods.py:963]: I agree that currently, it aligns with the FUGW API, where initialization adheres to the BaseMapping class:
```
class BaseMapping:
    def __init__(
        self,
        alpha=0.5,
        rho=1,
        eps=1e-2,
        reg_mode="joint",
        divergence="kl",
    ):
```
 Adopting a more sklearn-like approach could be advantageous, yet I haven't explored it in detail.

**Comment by pbarbarant** [comment]
OK, after digging into the sklearn API I see what you mean.

**Comment by bthirion** [comment]
We should probably discuss how we move forward with this. This could well be addressed in different PRs.

**Comment by emdupre** [comment]
I'll be out-of-office for a bit this summer, but if the timing lines up -- and if it would be useful to have me participate -- happy to have a discussion on the API !

**Comment by pbarbarant** [comment]
> I'll be out-of-office for a bit this summer, but if the timing lines up -- and if it would be useful to have me participate -- happy to have a discussion on the API !

Sure! 🚀

**Comment by pbarbarant** [comment]
> We should probably discuss how we move forward with this. This could well be addressed in different PRs.

The API design has been fixed.

**Review by bthirion — COMMENTED** [review]
The change LGTM.

**Comment by pbarbarant** [comment]
Feel free to merge whenever you want.

**Review by bthirion — COMMENTED** [review]
LGTM so far. Is this ready for merge ?

**Comment by pbarbarant** [comment]
> LGTM so far. Is this ready for merge ?

Yes it is

**Review by bthirion — APPROVED** [review]
LGTM.

**Comment by pbarbarant** [comment]
> LGTM.

I don't have maintainer access, can you merge it ?

**Pull request merged**